addons/mail/models/mail_message.py

```diff
@@ -503,15 +522,14 @@ def _get_forbidden_access(self, operation: str) -> api.Self:
 
         for model, docid_msgids in model_docid_msgids.items():
             documents = self.env[model].browse(docid_msgids)
-            if hasattr(documents, '_get_mail_message_access'):
-                doc_operation = documents._get_mail_message_access(docid_msgids, operation)  # why not giving model here?
-            else:
-                doc_operation = self.env['mail.thread']._get_mail_message_access(docid_msgids, operation, model_name=model)
-            doc_result = documents._check_access(doc_operation)
-            forbidden_doc_ids = set(doc_result[0]._ids) if doc_result else set()
-            for doc_id, msg_ids in docid_msgids.items():
-                if doc_id not in forbidden_doc_ids:
-                    for mid in msg_ids:
+            # returns {operation: res_ids to check}
+            # note that some ids may be filtered out if (e.g. group limitation, ...)
+            operation_res_ids = documents._group_mail_message_access(operation)
+            for doc_operation, doc_res_ids in operation_res_ids.items():
+                doc_result = documents.browse(doc_res_ids)._check_access(doc_operation)
```

Comment:
This `browse` in particular. It's say its best if it pre-fetched all `documents` rather than only those matching current operation.

If returning recordset is not practical, at least adding `with_prefetch(documents.ids)` here would probably be beneficial.